server/src/main/scala/com/cloudera/livy/server/LivyServer.scala

```diff
@@ -135,6 +167,54 @@ class LivyServer extends Logging {
     sys.props("livy.server.serverUrl") = _serverUrl.get
   }
 
+  def runKinit(keytab: String, principal: String): Boolean = {
+    val commands = Seq("kinit", "-kt", keytab, principal)
+    val proc = new ProcessBuilder(commands: _*).inheritIO().start()
+    proc.waitFor() match {
+      case 0 =>
+        debug("Ran kinit command successfully.")
+        kinitFailCount = 0
+        UserGroupInformation.getCurrentUser.reloginFromTicketCache()
+        true
+      case _ =>
+        warn("Fail to run kinit command.")
+        kinitFailCount += 1
+        false
+    }
+  }
+
+  def startKinitDaemon(keytab: String, principal: String): Unit = {
+    executor.schedule(
+      new Runnable() {
+        def run(): Unit = {
+          if (runKinit(keytab, principal)) {
+            // schedule another kinit run with a fixed delay.
+            executor.schedule(
+              new Runnable() {
+                def run(): Unit = {
+                  runKinit(keytab, principal)
+                }
+              }, livyConf.getInt(LAUNCH_KERBEROS_REFRESH_INTERVAL), TimeUnit.SECONDS)
+          } else {
+            // schedule another retry at once or fail the livy server if too many times kinit fail
+            val kinitFailThreshold = livyConf.getInt(KINIT_FAIL_THRESHOLD)
+            if (kinitFailCount >= kinitFailThreshold) {
+              error(s"Exit LivyServer after ${kinitFailThreshold} times failures running kinit.")
+              if (server.server.getState == "STARTED") {
+                stop()
+              } else {
+                sys.exit(1)
+              }
+            } else {
+              executor.submit(new Runnable() {
+                def run(): Unit = runKinit(keytab, principal)
```

Comment:
Similarly here; you want to schedule the task that does accounting and re-schedules a login task, not this custom task that only runs kinit and then doesn't do anything else.